src/main/java/de/zalando/ep/zalenium/proxy/DockerSeleniumRemoteProxy.java

```diff
@@ -451,30 +452,54 @@ public synchronized boolean isTestSessionLimitReached() {
         return getAmountOfExecutedTests() >= maxTestSessions;
     }
 
+    public boolean shutdownIfLimitReached() {
+        String currentName = configureThreadName();
+        boolean condition = isTestSessionLimitReached() && !isBusy();
+        if (condition) {
+            String msg = String.format("Exceded more than %s test sessions "
+                    + "in node %s", maxTestSessions, this.getContainerId());
+            LOGGER.debug(msg);
+            timeout(msg, ShutdownType.MAX_TEST_SESSIONS_REACHED);
+        }       
+        setThreadName(currentName);
+        return condition;
+    }
+
     public boolean shutdownIfIdle() {
         String currentName = configureThreadName();
-        boolean testIdle = isTestIdle();
-        boolean testSessionLimitReached = isTestSessionLimitReached();
-        boolean isShutdownIfIdle = testIdle || (testSessionLimitReached && !isBusy());
-        if (isShutdownIfIdle) {
-            LOGGER.debug("Proxy is idle.");
-            timeout("proxy being idle after test.", (testSessionLimitReached ?
-                    ShutdownType.MAX_TEST_SESSIONS_REACHED : ShutdownType.IDLE));
+        boolean condition = isTestIdleMoreThan(900) && !isCleaningUp() && !isBusy();
+        if (condition) {
+            String msg = String.format("Node %s has been idle for %s seconds. "
+                    + "Shutdown", this.getContainerId(), 900);
+            LOGGER.debug(msg);
+            timeout(msg, ShutdownType.IDLE);
         }
         setThreadName(currentName);
-        return isShutdownIfIdle;
+        return condition;
     }
 
+
     public boolean shutdownIfStale() {
         String currentName = configureThreadName();
-        if (isBusy() && isTestIdle() && !isCleaningUp()) {
-            LOGGER.debug("No test activity been recorded recently, proxy is stale.");
-            timeout("proxy being stuck | stale during a test.", ShutdownType.STALE);
+        boolean condition = isTestIdle() && !isCleaningUp() && isBusy();
+        if (condition) {
+            String msg = String.format("Test did not see a new command "
+                    + "for %s seconds. Timing out", getMaxTestIdleTimeSecs());
+            LOGGER.debug(msg);
+            timeout(msg, ShutdownType.STALE);
```

Comment:
Why don't we report the `MAX_TEST_SESSIONS_REACHED` shutdown type anymore? Could you please elaborate on the logical change here?

Reply:
It's reported in the shutdownIfLimitReached

Reply:
Yeah, but how can we keep the previous behaviour as well?